lib/Storage/Dropbox.php

```diff
@@ -0,0 +1,199 @@
+            $this->accessToken = $params['token'];
+            $this->root = isset($params['root']) ? $params['root'] : '/';
+
+            $app = new DropboxApp($this->clientId, $this->clientSecret, $this->accessToken);
+            $dropboxClient = new DropboxClient($app);
+
+            $this->adapter = new Adapter($dropboxClient);
+            $this->buildFlySystem($this->adapter);
+        } else {
+            throw new \Exception('Creating \OCA\Files_external_dropbox\Storage\Dropbox storage failed');
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getId() {
+        return 'dropbox_external::' . $this->clientId . '/' . $this->root;
+    }
+
+    public function file_exists($path) {
+        if ($path === '' || $path === '/' || $path === '.') {
+            return true;
+        }
+        return parent::file_exists($path);
+    }
+
+    protected function getLargest($arr, $default = 0) {
+        if (count($arr) === 0) {
+            return $default;
+        }
+        arsort($arr);
+        return array_values($arr)[0];
+    }
+
+    public function filemtime($path) {
+        if ($this->is_dir($path)) {
+            if ($path === '.' || $path === '') {
+                $path = "/";
+            }
+
+            if ($this->cacheFilemtime && isset($this->cacheFilemtime[$path])) {
+                return $this->cacheFilemtime[$path];
+            }
+
+            $arr = [];
+            $contents = $this->flysystem->listContents($path, true);
+            foreach ($contents as $c) {
+                $arr[] = $c['type'] === 'file' ? $c['timestamp'] : 0;
+            }
+            $mtime = $this->getLargest($arr);
+        } else {
+            if ($this->cacheFilemtime && isset($this->cacheFilemtime[$path])) {
+                return $this->cacheFilemtime[$path];
+            }
+            $mtime = parent::filemtime($path);
+        }
+        $this->cacheFilemtime[$path] = $mtime;
+        return $mtime;
+    }
+
+    public function stat($path) {
+        if ($path === '' || $path === '/' || $path === '.') {
+            return ['mtime' => 0];
+        }
+        return parent::stat($path);
+    }
+
+    public function getLatestCursor($path = '') {
+        try {
+            $dropbox = $this->adapter->getClient();
+            $resp = $dropbox->postToAPI('/files/list_folder/get_latest_cursor', ['path' => $path, 'recursive' => true, 'include_deleted' => true]);
+            $body = $resp->getDecodedBody();
+            if ($body) {
+                return $body['cursor'];
+            }
+        } catch (\Exception $e) {
+            // log it
+        }
+        return null;
+    }
+
+    public function isStorageUpdated($cursor) {
+        try {
+            $client = new \GuzzleHttp\Client();
+            $params = ['cursor' => $cursor, 'timeout' => 30];
+            $response = $client->post('https://notify.dropboxapi.com/2/files/list_folder/longpoll', ['json' => $params]);
```

Comment:
isn't longpoll a blocking operation ? If yes this would block other background jobs. ownCloud's background jobs are not designed to keep running forever.

Reply:
yes: https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder-longpoll

@Hemant-Mann please use the webhooks. If the instance is not reachable from the outside we have to fall back to the complete filescan.

Reply:
When looking at this code I was mostly expecting use of the delta API. Webhooks might be the best.

Reply:
@PVince81 Basically I am using longpoll to check whether any files have changed since the last time I scanned. There is no other way of knowing this because in v2 there is no delta API and if instance is not reachable this is the best way to do that.

If I scan the whole storage every time I run the background job it would definitely over load the API so longpoll tells me if any files/folders have changed. If yes then I find the changed folders and tell the scanner to scan only those folders instead of the whole storage.

Reply:
From my understanding of the docs we can just call `/delta/` and pass in the last cursor and get changes. If `!empty`, we can return true for `isStorageUpdated`

Reply:
and the make things quicker in the future we could use the webhooks to record the storage as updated in the background to avoid the request here

Reply:
Yes we could use webhooks in future no doubt
but `/delta` endpoint is in API v1 not in V2